### PR TITLE
Add subtask_price to TaskToCompute message

### LINKED ISSUE #157: Add subtask_price to the message TaskToCompute
And a proper comment like "total subtask price computed as price*subtask_timesout". And maybe some validation too, like subtask_price > 0.

### PR DESCRIPTION
**Pull request by kbeker**
Resolves #157

### CODE REVIEW

File: golem_messages/message/tasks.py [hunk 1]
```diff
@@ -107,6 +108,9 @@ def __init__(self, header: datastructures.MessageHeader = None,
         if self.concent_enabled is None:
             self.concent_enabled = True
 
+        if isinstance(self.subtask_price, int):
+            assert self.subtask_price > 0
```
Comment (shadeofblue) [inline golem_messages/message/tasks.py]: @kbeker is this requirement specified anywhere in the documentation? @lukasz-glen perhaps?

File: golem_messages/message/tasks.py [hunk 2]
```diff
@@ -96,6 +96,7 @@ class TaskToCompute(TaskMessageMixin, base.Message):
         'compute_task_def',
         'package_hash',
         'concent_enabled',
+        'subtask_price', # total subtask price computed as price*subtask_timesout
```
Comment (shadeofblue) [inline golem_messages/message/tasks.py]: typo -> should be `subtask_timeout` + I'd call it just `price`

File: golem_messages/message/tasks.py (hunk 1 above)
Comment (shadeofblue) [reply golem_messages/message/tasks.py]: @kbeker ah, just noticed @lukasz-glen 's comment in the issue

Comment (jiivan) [inline golem_messages/message/tasks.py]: 1. Probably `deserialize_slot()` is a better place for verification.
1. Please use `FieldError` ( see: https://github.com/golemfactory/golem-messages/blob/9edd8179f2c9a9aedc00879f505e056dddd9631a/golem_messages/validators.py#L6-L12 )
1. Assertions are disabled in optimized mode and shouldn't be an integral part of production logic: https://docs.python.org/3/library/constants.html#__debug__

File: golem_messages/message/tasks.py (hunk 2 above)
Comment (kbeker) [reply golem_messages/message/tasks.py]: @shadeofblue, Łukasz wrote on issue 
`And a proper comment like "total subtask price computed as price*subtask_timesout". And maybe some validation too, like subtask_price > 0.`
Shouldn't be `subtask_price` as he said?

File: golem_messages/message/tasks.py (hunk 1 above)
Comment (kbeker) [reply golem_messages/message/tasks.py]: Done, @jiivan please look at it one more time :)

File: tests/test_validators.py [hunk 3]
```diff
@@ -19,14 +20,23 @@ def validate(self, value):
             max_length=self.max_length,
         )
 
+    def validate_subtask_price_value(self, value):
+        return validators.validate_integer(
+            field_name=self.field_name,
+            value=value,
+        )
```
Comment (jiivan) [inline tests/test_validators.py]: I believe this is unrelated to `ValidateVarcharTestCase`

File: golem_messages/validators.py [hunk 4]
```diff
@@ -10,3 +10,12 @@ def validate_varchar(field_name, value, max_length):
             field=field_name,
             value=value,
         )
+
+
+def validate_integer(field_name, value):
+    if not ((isinstance(value, int)) or value is None):
```
Comment (jiivan) [inline golem_messages/validators.py]: Can `value` be `None` and still an instance of 'int'?

File: golem_messages/validators.py [hunk 5]
```diff
@@ -10,3 +10,12 @@ def validate_varchar(field_name, value, max_length):
...
+        raise exceptions.FieldError(
+            "Should be a integer field",
```
Comment (jiivan) [inline golem_messages/validators.py]: `an integer`

File: golem_messages/message/tasks.py (hunk 2 above)
Comment (shadeofblue) [reply golem_messages/message/tasks.py]: @kbeker nah, just make it `price` so as not to cause further proliferation of unneeded `subtask_` vs `task_` prefixes...

File: tests/test_validators.py (hunk 3 above)
Comment (shadeofblue) [reply tests/test_validators.py]: indeed, @kbeker please create a separate test case

File: tests/test_validators.py [hunk 6]
```diff
@@ -19,14 +20,23 @@ def validate(self, value):
...
+
     def test_value_too_long(self):
         with self.assertRaises(exceptions.FieldError):
             self.validate('*'*(self.max_length+1))
 
     def test_value_valid(self):
         for x in range(self.max_length+1):
             self.validate('*'*x)
+        self.validate_subtask_price_value(fake.pyint())
```
Comment (shadeofblue) [inline tests/test_validators.py]: this should be a separate test method ... though, generally, as mentioned above, it should be a separate test case anyway...

File: golem_messages/message/tasks.py (hunk 2 above)
Comment (shadeofblue) [reply golem_messages/message/tasks.py]: thank you

File: golem_messages/message/tasks.py [hunk 7]
```diff
@@ -123,6 +125,11 @@ def deserialize_slot(self, key, value):
         value = super().deserialize_slot(key, value)
         if key == 'compute_task_def':
             value = ComputeTaskDef(value)
+        if key == 'subtask_price':
```
Comment (shadeofblue) [inline golem_messages/message/tasks.py]: + please add two tests that load `TaskToCompute` with an integer, and a non-integer value in this field... as, sure, a test of the validator is fine, but I was more concerned about the `TaskToCompute` message itself...

**Comment by kbeker** [comment]
I have added all in one fixUp, @jiivan, @shadeofblue

**Comment by shadeofblue** [comment]
looks good now, imvho, though, I'd wait for @jiivan to have a look too

plus @lukasz-glen I'm unsure how all this relates to today's discussion re: price calculation in golem core (on #devs) -> maybe we should make sure once more that what we want to do here is indeed what should be done

File: golem_messages/message/tasks.py [hunk 8]
```diff
@@ -107,6 +108,10 @@ def __init__(self, header: datastructures.MessageHeader = None,
         if self.concent_enabled is None:
             self.concent_enabled = True
 
+        if self.price is None:
+            self.price = 0
```
[the remark this answers is not in the document]
Comment (kbeker) [reply golem_messages/message/tasks.py]: @jiivan Better to leave it None value?

[the remark this answers is not in the document]
Comment (jiivan) [reply golem_messages/message/tasks.py]: Yes, I believe so. `None` is default expected behaviour. `0` is custom.

[the remark this answers is not in the document]
Comment (kbeker) [reply golem_messages/message/tasks.py]: Ok. I've added to it also to validation that it could be None value.

File: golem_messages/validators.py [hunk 9]
```diff
@@ -13,7 +13,7 @@ def validate_varchar(field_name, value, max_length):
 
 
 def validate_integer(field_name, value):
-    if not isinstance(value, int):
+    if not isinstance(value, int) and value is not None:
```
Comment (shadeofblue) [inline golem_messages/validators.py]: @kbeker I'd just leave the previous check in place and treat `None` as illegal ... it should be the responsibility of the code that creates this message to ensure that it always sets the price... 

btw, that's e.g. why I had suggested, in one of my previous comments (https://github.com/golemfactory/golem-messages/pull/179#issuecomment-376936222) to add this property to the factory (here ->
https://github.com/golemfactory/golem-messages/blob/master/tests/factories.py#L36) so any tests that rely on factory-provided `TaskToCompute`, receive a well-formed message...

Comment (kbeker) [reply golem_messages/validators.py]: OK. I've added all, also I had to add few lines to `test_basic.py`
